### PR TITLE
framework(engine): refine the lifecycle of context used in resident goroutine

### LINKED ISSUE #7440: DM job can not be canceled
### What did you do?

```
$ curl -X POST  http://127.0.0.1:10240/api/v1/jobs/6c9c7dd1-bca7-4178-8068-2a3224d00110/cancel
{"id":"6c9c7dd1-bca7-4178-8068-2a3224d00110", "type":"DM", "state":"Canceling", "config":"dGFzay1tb2RlOiBhbGwKY29sbGF0aW9uX2NvbXBhdGlibGU6ICJzdHJpY3QiCnRhcmdldC1kYXRhYmFzZToKICBob3N0OiBob3N0LmRvY2tlci5pbnRlcm5hbAogIHBvcnQ6IDQwMDAKICB1c2VyOiByb290CiAgcGFzc3dvcmQ6ICcnCnVwc3RyZWFtczoKICAtIGRiLWNvbmZpZzoKICAgICAgaG9zdDogaG9zdC5kb2NrZXIuaW50ZXJuYWwKICAgICAgcG9ydDogMzMwNgogICAgICB1c2VyOiByb290CiAgICAgIHBhc3N3b3JkOiAnJwogICAgc291cmNlLWlkOiBteXNxbC0wMQogICAgYmxvY2stYWxsb3ctbGlzdDogZ2xvYmFsCiAgLSBkYi1jb25maWc6CiAgICAgIGhvc3Q6IGhvc3QuZG9ja2VyLmludGVybmFsCiAgICAgIHBvcnQ6IDMzMDcKICAgICAgdXNlcjogcm9vdAogICAgICBwYXNzd29yZDogJycKICAgIHNvdXJjZS1pZDogbXlzcWwtMDIKICAgIGJsb2NrLWFsbG93LWxpc3Q6IGdsb2JhbApibG9jay1hbGxvdy1saXN0OgogIGdsb2JhbDoKICAgIGRvLWRiczoKICAgIC0gInN5bmNfY29sbGF0aW9uKiIKICAgIC0gInRlc3RfcGFuaWMiCg==", "detail":"", "error":{"code":"None", "message":"", "not_leader":null}, "selectors":[]}% 
```

```
$ curl  http://127.0.0.1:10240/api/v1/jobs/6c9c7dd1-bca7-4178-8068-2a3224d00110/status
{
    "job_id": "6c9c7dd1-bca7-4178-8068-2a3224d00110",
    "task_status": {
        "mysql-01": {
            "expected_stage": "Running",
            "worker_id": "e102e019-b5ec-4b0c-853d-f486b913268d",
            "config_outdated": false,
            "status": {
                "error_message": "",
                "unit": "DMSyncTask",
                "stage": "Running",
                "result": null,
                "status": {
                    "binlogType": "remote",
                    "blockDDLOwner": "",
                    "blockingDDLs": [],
                    "conflictMsg": "",
                    "masterBinlog": "(ON.000001, 363)",
                    "masterBinlogGtid": "9a018a8a-5366-11ed-b0c2-0242c0a89008:1",
                    "recentTps": "0",
                    "secondsBehindMaster": "0",
                    "synced": true,
                    "syncerBinlog": "(ON.000001, 363)",
                    "syncerBinlogGtid": "9a018a8a-5366-11ed-b0c2-0242c0a89008:1",
                    "totalEvents": "2",
                    "totalTps": "0",
                    "unresolvedGroups": []
                }
            }
        },
        "mysql-02": {
            "expected_stage": "Running",
            "worker_id": "f4779aa5-87f7-400d-859f-5d6502bd4839",
            "config_outdated": false,
            "status": {
                "error_message": "",
                "unit": "DMSyncTask",
                "stage": "Running",
                "result": null,
                "status": {
                    "binlogType": "remote",
                    "blockDDLOwner": "",
                    "blockingDDLs": [],
                    "conflictMsg": "",
                    "masterBinlog": "(binlog.000002, 157)",
                    "masterBinlogGtid": "",
                    "recentTps": "0",
                    "secondsBehindMaster": "0",
                    "synced": true,
                    "syncerBinlog": "(binlog.000002, 157)",
                    "syncerBinlogGtid": "",
                    "totalEvents": "2",
                    "totalTps": "0",
                    "unresolvedGroups": []
                }
            }
        }
    }
}%                       
```

### What did you expect to see?

_No response_

### What did you see instead?

add diff

```diff
diff --git a/engine/jobmaster/dm/dm_jobmaster.go b/engine/jobmaster/dm/dm_jobmas
ter.go
index e412c26a1..fa6acdafe 100644
--- a/engine/jobmaster/dm/dm_jobmaster.go
+++ b/engine/jobmaster/dm/dm_jobmaster.go
@@ -398,6 +398,7 @@ func (jm *JobMaster) cancel(ctx context.Context, code frameModel.WorkerState) er
 
        if err := jm.taskManager.OperateTask(ctx, dmpkg.Deleting, nil, nil); err != nil {
                // would not recover again
+               jm.Logger().Warn("failed to mark task deleting", zap.Error(err))
                return jm.Exit(ctx, framework.ExitReasonCanceled, err, detail)
        }
        // wait all worker exit
@@ -405,9 +406,11 @@ func (jm *JobMaster) cancel(ctx context.Context, code frameModel.WorkerState) er
        for {
                select {
                case <-ctx.Done():
+                       jm.Logger().Warn("cancel context is timeout", zap.Error(ctx.Err()))
                        return jm.Exit(ctx, framework.ExitReasonCanceled, ctx.Err(), detail)
                case <-time.After(time.Second):
                        if jm.workerManager.allTombStone() {
+                               jm.Logger().Info("all worker are offline, will exit")
                                return jm.Exit(ctx, framework.WorkerStateToExitReason(status.State), err, detail)
                        }
                        jm.workerManager.SetNextCheckTime(time.Now())
```

executor

```
[2022/10/24 08:01:33.259 +00:00] [INFO] [dm_jobmaster.go:273] ["on cancel job master"] [job_id=6c9c7dd1-bca7-4178-8068-2a3224d00110]
[2022/10/24 08:01:33.261 +00:00] [DEBUG] [message_agent.go:338] ["send request"] [job_id=6c9c7dd1-bca7-4178-8068-2a3224d00110] [component=message-agent] [client-id=mysql-02] [command=QueryStatus] [req="{\"Task\":\"mysql-02\"}"]
[2022/10/24 08:01:33.261 +00:00] [DEBUG] [message_agent.go:338] ["send request"] [job_id=6c9c7dd1-bca7-4178-8068-2a3224d00110] [component=message-agent] [client-id=mysql-01] [command=QueryStatus] [req="{\"Task\":\"mysql-01\"}"]
[2022/10/24 08:01:33.261 +00:00] [DEBUG] [client.go:267] ["Sending Message"] [topic=DM/36633963376464312d626361372d343137382d383036382d326133323234643030313130/6d7973716c2d3031] [seq=4]
[2022/10/24 08:01:33.261 +00:00] [DEBUG] [client.go:267] ["Sending Message"] [topic=DM/36633963376464312d626361372d343137382d383036382d326133323234643030313130/6d7973716c2d3032] [seq=4]
[2022/10/24 08:01:33.261 +00:00] [INFO] [task_manager.go:68] ["operate task"] [job_id=6c9c7dd1-bca7-4178-8068-2a3224d00110] [component=task_manager] [op=Deleting] [tasks="[]"]
[2022/10/24 08:01:33.265 +00:00] [ERROR] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=41.725µs] [sql="INSERT INTO `)_meta_kvs` (`created_at`,`updated_at`,`meta_key`,`meta_value`,`job_id`) VALUES ('2022-10-24 08:01:33.265','2022-10-24 08:01:33.265','/data-flow/dm/job','<binary>','6c9c7dd1-bca7-4178-8068-2a3224d00110') ON DUPLICATE KEY UPDATE `updated_at`='2022-10-24 08:01:33.265',`meta_key`=VALUES(`meta_key`),`meta_value`=VALUES(`meta_value`),`job_id`=VALUES(`job_id`)"] [affected-rows=0] [error="context canceled"]
[2022/10/24 08:01:33.265 +00:00] [WARN] [dm_jobmaster.go:401] ["failed to mark task deleting"] [job_id=6c9c7dd1-bca7-4178-8068-2a3224d00110] [error="[DFLOW:ErrMetaOpFail]meta operation fail: context canceled"] [errorVerbose="[DFLOW:ErrMetaOpFail]meta operation fail: context canceled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20220729040631-518f63d66278/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20220729040631-518f63d66278/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/engine/jobmaster/dm/metadata.(*TomlStore).Put\n\tgithub.com/pingcap/tiflow/engine/jobmaster/dm/metadata/state.go:99\ngithub.com/pingcap/tiflow/engine/jobmaster/dm/metadata.(*JobStore).MarkDeleting\n\tgithub.com/pingcap/tiflow/engine/jobmaster/dm/metadata/job.go:239\ngithub.com/pingcap/tiflow/engine/jobmaster/dm.(*TaskManager).OperateTask\n\tgithub.com/pingcap/tiflow/engine/jobmaster/dm/task_manager.go:83\ngithub.com/pingcap/tiflow/engine/jobmaster/dm.(*JobMaster).cancel\n\tgithub.com/pingcap/tiflow/engine/jobmaster/dm/dm_jobmaster.go:399\ngithub.com/pingcap/tiflow/engine/jobmaster/dm.(*JobMaster).OnCancel\n\tgithub.com/pingcap/tiflow/engine/jobmaster/dm/dm_jobmaster.go:274\ngithub.com/pingcap/tiflow/engine/framework.(*jobMasterImplAsWorkerImpl).OnMasterMessage\n\tgithub.com/pingcap/tiflow/engine/framework/base_jobmaster.go:421\ngithub.com/pingcap/tiflow/engine/framework.(*DefaultBaseWorker).doPreInit.func3\n\tgithub.com/pingcap/tiflow/engine/framework/worker.go:338\ngithub.com/pingcap/tiflow/engine/framework.(*MessageRouter).routeMessage.func1\n\tgithub.com/pingcap/tiflow/engine/framework/message_router.go:104\ngithub.com/pingcap/tiflow/pkg/workerpool.(*asyncWorker).run\n\tgithub.com/pingcap/tiflow/pkg/workerpool/async_pool_impl.go:166\ngithub.com/pingcap/tiflow/pkg/workerpool.(*defaultAsyncPoolImpl).Run.func2\n\tgithub.com/pingcap/tiflow/pkg/workerpool/async_pool_impl.go:114\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1594"]
[2022/10/24 08:01:33.265 +00:00] [ERROR] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=26.295µs] [sql="UPDATE `master_meta` SET `detail`='{\"job_id\":\"6c9c7dd1-bca7-4178-8068-2a3224d00110\",\"task_status\":{\"mysql-01\":{\"expected_stage\":\"Running\",\"worker_id\":\"e102e019-b5ec-4b0c-853d-f486b913268d\",\"config_outdated\":false,\"status\":{\"error_message\":\"context canceled\",\"unit\":\"\",\"stage\":\"\",\"result\":null,\"status\":null}},\"mysql-02\":{\"expected_stage\":\"Running\",\"worker_id\":\"f4779aa5-87f7-400d-859f-5d6502bd4839\",\"config_outdated\":false,\"status\":{\"error_message\":\"context canceled\",\"unit\":\"\",\"stage\":\"\",\"result\":null,\"status\":null}}}}',`error_message`='[DFLOW:ErrMetaOpFail]meta operation fail: context canceled',`state`='4',`updated_at`='2022-10-24 08:01:33.266' WHERE id = '6c9c7dd1-bca7-4178-8068-2a3224d00110' AND `master_meta`.`deleted` IS NULL"] [affected-rows=0] [error="context canceled"]
[2022/10/24 08:01:33.299 +00:00] [DEBUG] [server.go:726] ["received packet"] [streamHandle=server-executor-3aff82f5] [numEntries=2]
[2022/10/24 08:01:33.302 +00:00] [DEBUG] [message_agent.go:359] ["on message"] [job_id=6c9c7dd1-bca7-4178-8068-2a3224d00110] [worker_id=e102e019-b5ec-4b0c-853d-f486b913268d] [component=message-agent] [topic=DM/36633963376464312d626361372d343137382d383036382d326133323234643030313130/6d7973716c2d3031] [msg="{\"ID\":8,\"Type\":2,\"Command\":\"QueryStatus\",\"Payload\":{\"Task\":\"mysql-01\"}}"]
```

master

```
[2022/10/24 08:01:33.175 +00:00] [INFO] [server.go:199] [payload="id:\"6c9c7dd1-bca7-4178-8068-2a3224d00110\""] [request=CancelJob]
[2022/10/24 08:01:33.177 +00:00] [DEBUG] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=1.399053ms] [sql="SELECT * FROM `master_meta` WHERE id = '6c9c7dd1-bca7-4178-8068-2a3224d00110' AND `master_meta`.`deleted` IS NULL ORDER BY `master_meta`.`seq_id` LIMIT 1"] [affected-rows=1]
[2022/10/24 08:01:33.178 +00:00] [DEBUG] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=668.637µs] [sql="SELECT count(*) FROM `job_ops` WHERE job_id = '6c9c7dd1-bca7-4178-8068-2a3224d00110'"] [affected-rows=1]
[2022/10/24 08:01:33.179 +00:00] [DEBUG] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=815.37µs] [sql="INSERT INTO `job_ops` (`created_at`,`updated_at`,`op`,`job_id`) VALUES ('2022-10-24 08:01:33.179','2022-10-24 08:01:33.179','2','6c9c7dd1-bca7-4178-8068-2a3224d00110') ON DUPLICATE KEY UPDATE `updated_at`=VALUES(`updated_at`),`op`=VALUES(`op`),`job_id`=VALUES(`job_id`)"] [affected-rows=1]
[2022/10/24 08:01:33.181 +00:00] [DEBUG] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=233.921µs] [sql="SELECT * FROM `job_ops` WHERE op = '2'"] [affected-rows=1]
[2022/10/24 08:01:33.181 +00:00] [DEBUG] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=250.516µs] [sql="SELECT * FROM `master_meta` WHERE id = '6c9c7dd1-bca7-4178-8068-2a3224d00110' AND `master_meta`.`deleted` IS NULL ORDER BY `master_meta`.`seq_id` LIMIT 1"] [affected-rows=1]
[2022/10/24 08:01:33.181 +00:00] [DEBUG] [client.go:267] ["Sending Message"] [topic=worker-status-change-req-dataflow-engine-job-manager-6c9c7dd1-bca7-4178-8068-2a3224d00110] [seq=1]
[2022/10/24 08:01:33.392 +00:00] [DEBUG] [server.go:1129] ["request handled successfully"] [method=/enginepb.Discovery/Heartbeat] [request="executor_id:\"server-executor-3aff82f5\"  status:1  timestamp:1666598493  ttl:23000"] [response="leader:\"server-master:10240\"  addrs:\"server-master:10240\""]
[2022/10/24 08:01:33.892 +00:00] [DEBUG] [server.go:1129] ["request handled successfully"] [method=/enginepb.Discovery/Heartbeat] [request="executor_id:\"server-executor-3aff82f5\"  status:1  timestamp:1666598493  ttl:23000"] [response="leader:\"server-master:10240\"  addrs:\"server-master:10240\""]
[2022/10/24 08:01:34.393 +00:00] [DEBUG] [server.go:1129] ["request handled successfully"] [method=/enginepb.Discovery/Heartbeat] [request="executor_id:\"server-executor-3aff82f5\"  status:1  timestamp:1666598494  ttl:23000"] [response="leader:\"server-master:10240\"  addrs:\"server-master:10240\""]
[2022/10/24 08:01:34.735 +00:00] [DEBUG] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=1.438632ms] [sql="SELECT * FROM `job_ops` WHERE op = '2'"] [affected-rows=1]
[2022/10/24 08:01:34.737 +00:00] [DEBUG] [callbacks.go:134] ["trace log"] [component=gorm] [elapsed=1.236756ms] [sql="SELECT * FROM `master_meta` WHERE id = '6c9c7dd1-bca7-4178-8068-2a3224d00110' AND `master_meta`.`deleted` IS NULL ORDER BY `master_meta`.`seq_id` LIMIT 1"] [affected-rows=1]
[2022/10/24 08:01:34.737 +00:00] [DEBUG] [client.go:267] ["Sending Message"] [topic=worker-status-change-req-dataflow-engine-job-manager-6c9c7dd1-bca7-4178-8068-2a3224d00110] [seq=2]
[2022/10/24 08:01:34.793 +00:00] [DEBUG] [elector.go:117] [renew] [cost=5.970652ms] []
```
```

### Versions of the cluster

Dataflow Engine version (run `tiflow version`):

```console
(paste Dataflow Engine version here)
```

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve?

Issue Number: close #7440

### What is changed and how it works?

The `context` used in `Init` function of framework should not be canceled after `Init` function returns, since this context will be registered to message handler. Only cancel the context when first error is met.

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by lance6716 — APPROVED** [review]

**Comment by gozssky** [comment]
> The context used in Init function of framework should not be canceled after Init function returns, since this context will be registered to message handler. Only cancel the context when first error is met.

Canceling the context after `Init` function returns is intentional. This is to make sure any implementation does not hold the context after the function returns. Otherwise, it's hard for us to manage the lifecycle of the context. 

On the other hand, `d.errCenter.WithCancelOnFirstError(ctx)` will register the new context to parent context's map. If there are always no errors and Init is called multiple times, a memory leak will occur. This is similar to `context.WithCancel`.

**Comment by amyangfei** [comment]
> Canceling the context after Init function returns is intentional. This is to make sure any implementation does not hold the context after the function returns. Otherwise, it's hard for us to manage the lifecycle of the context.

The real case is that we have some background goroutine will hold the context after `Init` returns, including `goroutine` starts from business logic in `Impl.InitImpl` function, and message handler callback. In current design seems there is no better to avoid this.

> On the other hand, d.errCenter.WithCancelOnFirstError(ctx) will register the new context to parent context's map. If there are always no errors and Init is called multiple times, a memory leak will occur. This is similar to context.WithCancel.

There are two scenarios for the error center, if error happens, all children context will be canceled. On the other hand if no error happens, `Init` won't be called multiple times, which is guaranteed in the upper runtime layer. IMO there is no memory leak. The usage of `d.errCenter.WithCancelOnFirstError(ctx)` is more like

```go
func entrance() {
        cancel := worker(ctx)
        // meets error, cancel worker
        cancel()
}

func worker(ctx context.Context) context.CancelFunc {
        ctx, cancel := context.WithCancel(ctx)
        go func() { fn(ctx) }()
        return cancel
}
```

**Comment by gozssky** [comment]
> The real case is that some background goroutines will hold or use the context after Init returns, including goroutine starts from business logic in Impl.InitImpl function, and message handler callback. In current design seems there is no better to avoid this.

We also have `InitImpl.CloseImpl` function, maybe we can call `InitImpl.CloseImpl` explicitly when there is any error occurs. Most methods of the `WorkerImpl` interface have a context parameter, but it's not clear which context may be held by the method implementation. If we allow method implementation to hold the context after returns, it is very easy for the caller to accidentally cancel the context.

**Comment by gozssky** [comment]
> Init won't be called multiple times, which is guaranteed in the upper runtime layer. IMO there is no memory leak

What if `Init` fails or the worker exits?

**Comment by amyangfei** [comment]
> > The real case is that some background goroutines will hold or use the context after Init returns, including goroutine starts from business logic in Impl.InitImpl function, and message handler callback. In current design seems there is no better to avoid this.
> 
> We also have `InitImpl.CloseImpl` function, maybe we can call `InitImpl.CloseImpl` explicitly when there is any error occurs. Most methods of the `WorkerImpl` interface have a context parameter, but it's not clear which context may be held by the method implementation. If we allow method implementation to hold the context after returns, it is very easy for the caller to accidentally cancel the context.

Yes, it is dangerous to leave the context cancel outside, and must be careful with the code. I have reverted the change and re-added the `defer cancel()` call.
Just as a workaround for the context used in resident, use a context derived from background context and cancel it in the `worker.Close()`. PTAL @gozssky

**Review by CharlesCheung96 — APPROVED** [review]

File: engine/framework/worker.go [hunk 1]
```diff
@@ -301,8 +302,9 @@ func (w *DefaultBaseWorker) doPreInit(ctx context.Context) (retErr error) {
 			retErr = frameErrors.FailFast(retErr)
 		}
 	}()
-	// TODO refine this part
-	poolCtx, cancelPool := context.WithCancel(context.TODO())
+	// poolCtx will be hold in background goroutines, and it won't be canceled
```
Comment (gozssky) [inline engine/framework/worker.go:305]:
```suggestion
	// poolCtx will be held in background goroutines, and it won't be canceled
```

**Review by gozssky — APPROVED** [review]
The PR title is outdated.

**Pull request merged**